hazelcast/src/jepsen/hazelcast.clj

```diff
@@ -257,55 +329,235 @@
+  (do
+    (try (Thread/sleep LOGGING_SLEEP_DURATION_MS) (catch Exception _))
+    (info (str "_" clientName "_release_fail " (:value op)))))
+
+(defn log-release-maybe
+  [clientName op exception]
+  (do
+    (try (Thread/sleep LOGGING_SLEEP_DURATION_MS) (catch Exception _))
+    (info (str "_" clientName "_release_maybe " (:value op) " exception: " (.getMessage exception)))))
+
+(defn invoke-acquire [clientName lock op]
+  (let [fence (.tryLockAndGetFence lock 5000 TimeUnit/MILLISECONDS)]
+    (if (not= fence INVALID_FENCE)
+      (do
+        (log-acquire-ok clientName op)
+        (assoc op :type :ok :value {:client clientName :fence fence :uid (:value op)})
+        )
+      (do
+        (log-acquire-fail clientName op)
+        (assoc op :type :fail)
+        )
+      )
+    )
+  )
+
+(defn fenced-lock-client
+  ([lock-name] (fenced-lock-client nil nil lock-name))
+  ([conn lock lock-name]
+   (reify client/Client
+     (setup! [_ test node]
+       (let [conn (connect node)]
+         (fenced-lock-client conn (.getLock (.getCPSubsystem conn) lock-name) lock-name)
+         )
+       )
+
+     (invoke! [this test op]
+       (let [clientName (.getName conn)]
+         (try
+           (info (str " " CLIENT_ID_SEPARATOR " " clientName " " CLIENT_ID_SEPARATOR " " op))
+           (case (:f op)
+             :acquire (invoke-acquire clientName lock op)
+             :release (do
+                        (.unlock lock)
+                        (log-release-ok clientName op)
+                        (assoc op :type :ok :value {:client clientName :uid (:value op)})
+                        )
+             )
+           (catch IllegalMonitorStateException e
+             (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+             (assoc op :type :fail, :error :not-lock-owner)
+             )
+           (catch IOException e
+             (condp re-find (.getMessage e)
+               ; This indicates that the Hazelcast client doesn't have a remote
+               ; peer available, and that the message was never sent.
+               #"Packet is not send to owner address"
+               (do
+                 (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+                 (assoc op :type :fail, :error :client-down)
+                 )
+               (do
+                 (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+                 (assoc op :type :info, :error :io-exception)
+                 )
+               )
+             )
+           (catch Exception e
+             (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+             (assoc op :type :info, :error :exception)
+             )
+           )
+         )
+       )
+
+     (teardown! [this test]
+       (try
+         (.terminate (.getLifecycleService conn))
```

Comment:
In the other clients, you call `(.shutdown conn)`--do you also want to call `.shutdown` here? Or `(.terminate (.getLifecycleService conn))` in other clients?